docs/03-code-internals/19-service-objects.md

```diff
@@ -0,0 +1,602 @@
+```
+
+## Actions
+
+When a step starts becoming too complex, like it has too many branching statements for example, then it’s time to extract all that logic to a dedicated class. That logic could live in a model for instance, but when in doubt, just create a new action.
+An action is just a small class that responds to a `.call` method by convention. What happens inside is up to you. The idea, however, is to execute an action (hence the name) with minimal overhead. It means an action should not validate data, for example. It should be called with valid objects only, thus being able to work with them right away. It also means that an action should not fail. You can think of an action as a bare-bones service without all the bells and whistles. Also, an action can be reused by different services.
+
+Here again, it’s relatively simple to create a new action. Let’s take as an example our `log` step:
+```ruby
+class User::Action::LogUsernameChange < Service::ActionBase
+  option :actor
+  option :user
+
+  def call
+    StaffActionLogger.new(actor).log_username_change(
+      user,
+      user.username_before_last_save,
+      user.username,
+    )
+  end
+end
+```
+Of course, this is a very basic example, you can do more complex things in an action. A real-world example can be found in [`User::Action::TriggerPostAction`](https://github.com/discourse/discourse/blob/main/app/services/user/action/trigger_post_action.rb).
+`Service::ActionBase` comes with [`Dry::Initializer`](https://dry-rb.org/gems/dry-initializer) which provides a nice mini-DSL:
+- use `param :my_arg` to declare a required argument named `my_arg`.
+- use `option :my_arg` to declare a required named argument named `my_arg`.
+- use `optional: true` to declare the argument optional (for instance, `option :my_arg, optional: true`).
+
+You should not need anything more than that to work with an action, but if you want to use some advanced features of `dry-initializer` (like coercion), just [take a look at their docs](https://dry-rb.org/gems/dry-initializer).
+
+Some rules to keep in mind when writing an action:
+- it has to inherit from `Service::ActionBase`.
+- it has to define one method: `#call`.
+- prefer `option` over `param` to define arguments, as it’s a bit more self-documenting on the caller side. It also allows you to use the [hash shorthand syntax](https://bugs.ruby-lang.org/issues/17292).
+- by convention, it should be namespaced under its concept followed by the `Action` namespace: for our current example, it means `User::Action::` which maps to `app/services/user/action/` on the filesystem.
+
+> 💡 To keep things short and clear, feel free to use [`delegate`](https://api.rubyonrails.org/classes/Module.html#method-i-delegate) extensively.
+
+Then, when you want to use it in a service, just write your step like this:
+```ruby
+def log(guardian:, user:)
+  User::Action::LogUsernameChange.call(actor: guardian.user, user:)
+end
+```
+
+# Best practices and guidelines
```

Comment:
Should we add something in here (or elsewhere) about calling services from services, which we want people to avoid?

Reply:
Yeah, I was wondering about that, we discussed it a bit with @jjaffeux during the meetup and now that we can declare an outcome block when calling the service with `.call`, maybe that’s not so problematic anymore.

Inside a step, we can do something like:
```ruby
def my_step
  MyOtherService.call(param1:, param2:) do
    on_success { do_something }
    on_failure { handle_failure }
    …
  end
end
```
In the various outcome matchers, we can put something in the caller `context` for instance, or even call `fail!`.
That’s still something to think about though :grin: